content/2021-06-07-quarkus-func-payload-mapper.md

```diff
@@ -0,0 +1,308 @@
+            }
+        },
+        "temp": {
+            "properties": {
+                "value": 23.0
+            }
+        }
+    }
+}
+```
+
+This use case is quite common if we want to look at the Drogue cloud as an IoT integration platform. Eclipse Ditto is just one of the services where our telemetry data will end up, but you can imagine various others like time series databases, data lakes, etc. all with their own payload formats.
+
+If you take a look at these kind of services we can see that they are simple, stateless services with well defined input and output. And we have to be able to scale them up and down depending on the data volume. That sounds like a perfect serverless use case.
+
+All this led to the idea of using Func with Quarkus runtime to implement one such service and see what's the developer experience in doing so. So, let's see.
+
+## Cloud native Ditto
+
+Before we dive into implementing the converter service, let's quickly recap Ditto and how it fits into our stack. Ditto provides an HTTP API which can be used to manage your things. In order to start, you need to first create your "thing" using the HTTP API:
+
+```shell
+echo "{}" | http --auth ditto:ditto PUT http://$TWIN_API/api/2/things/app_id:simple-thing
+```
+
+Beyond HTTP API, Ditto provides a Cloud Events API, which allow us to issue commands (from the example above) as well formed cloud events, like:
+
+```shell
+cat modify.json |  http -v --auth ditto:ditto http://$TWIN_API/api/2/cloudevents \
+  Content-Type:application/json \
+  Ce-Id:9bc19ad9-1c37-4041-b488-5b622e63bbfa \
+  Ce-Source:drogue://app%5Fid/simple%2Dthing \
+  Ce-Subject:foo \
+  Ce-Type:io.drogue.event.v1 \
+  Ce-Specversion:1.0 \
+  Ce-Application:app \
+  Ce-Device:device
+```
+
+And we can make sure that our changes took place:
+
+```shell
+http --auth ditto:ditto http://$TWIN_API/api/2/things/app_id:simple-thing
+HTTP/1.1 200 OK
+...
+etag: "rev:2"
+response-required: false
+version: 2
+
+{
+    "features": {
+        "hum": {
+            "properties": {
+                "value": "52.0"
+            }
+        },
+        "temp": {
+            "properties": {
+                "value": 22.5
+            }
+        }
+    },
+    "policyId": "app_id:simple-thing",
+    "thingId": "app_id:simple-thing"
+}
+```
+
+Why is this important? Well, it allow us to integrate it into the Knative workflows. Take a look at the following example:
+
+~~~yaml
+apiVersion: sources.knative.dev/v1alpha1
+kind: KafkaSource
+metadata:
+  name: ditto-converter-kafka-source
+  labels:
+    app.kubernetes.io/part-of: digital-twin
+spec:
+  consumerGroup: ditto-converter
+  bootstrapServers:
+    - kafka-eventing-kafka-bootstrap.knative-eventing.svc:9092
+  topics:
+    - knative-messaging-kafka.drogue-iot.iot-channel
+  sink:
+    ref:
+      apiVersion: flows.knative.dev/v1
+      kind: Sequence
+      name: ditto-converter
+---
+apiVersion: flows.knative.dev/v1
+kind: Sequence
```

Comment:
It's not completely clear to me why the Sequence is needed here. And why both the KafkaSource and KafkaChannel are referenced. But maybe the intended audience for the article won't care about those details?

Reply:
This is how we used it in the original twin post: https://blog.drogue.io/digital-twins/ ... Having a sequence is a nice way to have a response from a service get forwarded to the other service in the ends. I want to improve this in the future and add triggers and filter for event types, but I thought to do that in the follow-up